### PR TITLE
Checkpointer

### PR DESCRIPTION
**Pull request by dlwh**
here's a cut at checkpointing we outlined last week:

It saves a checkpoint at least every `save_interval` seconds (default 6 hours), and supports saving checkpoints based on a number of steps (which can vary over the course of the run, a la Mistral). the "save_interval" triggered checkpoints only keep the most recent.

I'm a bad person and didn't really write unit tests for this (I guess I could do DI for the checkpointer...) I'll try to write some tonight I think.

### CODE REVIEW

File: src/levanter/config.py [hunk 1]
```diff
@@ -348,5 +379,34 @@ def name(dtype):
     pyrallis.decode.register(jmp.Policy, lambda policy_str: jmp.get_policy(policy_str))
     pyrallis.encode.register(jmp.Policy, policy_encode)
 
+    def parse_timedelta(td_str):
+        import pytimeparse
+
+        return timedelta(seconds=pytimeparse.parse(td_str))
+
+    def encode_timedelta(td: timedelta):
```
Comment (dlwh) [inline src/levanter/config.py:387]: i can't believe there's no builtin or easily pip installable library for this?

**Review by yifanmai — CHANGES_REQUESTED** [review]

File: src/levanter/checkpoint.py [hunk 2]
```diff
@@ -17,6 +19,99 @@
...
+            path = self.path
+        return load_checkpoint(model, training_state, path, discover_latest=discover_latest)
+
+    def load_model(self, model, path: Optional[str] = None, *, discover_latest: bool = True):
+        if path is None:
+            path = self.path
+        return load_checkpoint(model, None, path, discover_latest=discover_latest)
+
+    def __call__(self, info, force: bool = False):
+        self.on_step(info)
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/checkpoint.py:57]: oops, thanks. i added that late in the game

File: src/levanter/checkpoint.py [hunk 3]
```diff
@@ -17,6 +19,99 @@
...
+class Checkpointer:
+    """A checkpointer class that saves checkpoints with two different, but overlapping policies: time and step."""
+
+    path: str
+    save_interval: datetime.timedelta
+    keep: Sequence[CheckpointInterval] = dataclasses.field(default_factory=lambda: [CheckpointInterval(every=1000)])
+
+    _last_temporary_checkpoint: Optional[str] = None
+
+    def __init__(self, path: str, save_interval: datetime.timedelta, keep: Sequence[CheckpointInterval]):
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/checkpoint.py:37]: it's not so much default_path as base_path. I'll do that?

[the remark this answers is not in the document]
Comment (yifanmai) [reply src/levanter/checkpoint.py:37]: `base_path` sounds good.

File: src/levanter/checkpoint.py [hunk 4]
```diff
@@ -17,6 +19,99 @@
...
+        if path is None:
+            path = self.path
+        return load_checkpoint(model, training_state, path, discover_latest=discover_latest)
+
+    def load_model(self, model, path: Optional[str] = None, *, discover_latest: bool = True):
+        if path is None:
+            path = self.path
+        return load_checkpoint(model, None, path, discover_latest=discover_latest)
+
+    def __call__(self, info, force: bool = False):
```
[the remark this answers is not in the document]
Comment (yifanmai) [reply src/levanter/checkpoint.py:56]: I think a method should work as a callable i.e. `engine.add_hook(checkpointer.some_method, every=1)` should still work

File: src/levanter/config.py [hunk 5]
```diff
@@ -348,5 +379,34 @@ def name(dtype):
     pyrallis.decode.register(jmp.Policy, lambda policy_str: jmp.get_policy(policy_str))
     pyrallis.encode.register(jmp.Policy, policy_encode)
 
+    def parse_timedelta(td_str):
+        import pytimeparse
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/config.py:383]: i don't remember why i did this

File: src/levanter/config.py [hunk 6]
```diff
@@ -150,6 +153,39 @@ def _infer_experiment_git_root() -> Optional[str]:
...
+class CheckpointerConfig:
+    base_path: str = "checkpoints/"
+    save_interval: timedelta = timedelta(hours=6)
+    # TODO: I'd like to write this, but it's not supported by pyrallis
+    # keep: List[CheckpointInterval] = field(default_factory=lambda: [CheckpointInterval(every=1000)])
+    keep: List[dict] = field(
+        default_factory=lambda: [dict(every=1000)]
+    )  # list of dicts with two keys: every and until
+
+    def create(self, run_name) -> Checkpointer:
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/config.py:166]: i saw this as a builder pattern where having a build/create method on the builder/config is pretty normal?

**Comment by dlwh** [comment]
yeah i realized that after i replied... I'm still noobish at python
sometimes

File: src/levanter/config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/config.py:166]: (i left this as-is for now. if you feel strongly I'll change, but I prefer to keep the run_name-based knot-tying in the config classes)

[the remark this answers is not in the document]
Comment (yifanmai) [reply src/levanter/config.py:166]: Sounds good to me.

**Pull request merged**